### PR TITLE
Added backup support for `google_filestore_instance`

### PR DESCRIPTION
**Pull request by googlerjk**
Added resource `google_filestore_backup`
 Added `sourceBackup` field to `google_filestore_instance` resource
fixes https://github.com/hashicorp/terraform-provider-google/issues/7655

If this PR is for Terraform, I acknowledge that I have:

- [x] Searched through the [issue tracker](https://github.com/hashicorp/terraform-provider-google/issues) for an open issue that this either resolves or contributes to, commented on it to claim it, and written "fixes {url}" or "part of {url}" in this PR description. If there were no relevant open issues, I opened one and commented that I would like to work on it (not necessary for very small changes).
- [x] [Generated Terraform](https://github.com/GoogleCloudPlatform/magic-modules#generating-the-terraform-providers), and ran `make test` and `make lint` to ensure it passes unit and linter tests.
- [x] Ensured that all new fields I added that can be set by a user appear in at least one [example](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/templates/terraform/examples) (for generated resources) or [third_party test](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/third_party/terraform/tests) (for handwritten resources or update tests).
- [x] [Ran](https://github.com/hashicorp/terraform-provider-google/blob/main/.github/CONTRIBUTING.md#tests) relevant acceptance tests (If the acceptance tests do not yet pass or you are unable to run them, please let your reviewer know).
- [x] Read the [Release Notes Guide](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/.ci/RELEASE_NOTES_GUIDE.md) before writing my release note below.

**Release Note Template for Downstream PRs (will be copied)**

```release-note:new-resource
`google_filestore_backup`
```

### CODE REVIEW

File: mmv1/products/filestore/api.yaml [hunk 1]
```diff
@@ -188,6 +188,14 @@ objects:
                       An integer representing the anonymous group id with a default value of 65534.
                       Anon_gid may only be set with squashMode of ROOT_SQUASH. An error will be returned
                       if this field is specified for other squashMode settings.
+            - !ruby/object:Api::Type::String
+              name: 'sourceBackup'
+              description: |
+                The resource name of the backup, in the format
+                projects/{projectId}/locations/{locationId}/backups/{backupId},
+                that this file share has been restored from.
+              required: false
```
Comment (c2thorn) [inline mmv1/products/filestore/api.yaml:197]:
```suggestion
```
the omission of a `required` field also means it is optional.

File: mmv1/products/redis/api.yaml [hunk 2]
```diff
@@ -160,7 +160,7 @@ objects:
               - :RDB
           - !ruby/object:Api::Type::Enum
             name: 'rdbSnapshotPeriod'
-            required: true
+            required: false
```
Comment (c2thorn) [inline mmv1/products/redis/api.yaml:163]:
```suggestion
```
see above comment about `required`. Also why is this in this PR?

File: mmv1/third_party/validator/tests/data/example_filestore_backup.json [hunk 3]
```diff
@@ -0,0 +1,21 @@
+[
```
Comment (c2thorn) [inline mmv1/third_party/validator/tests/data/example_filestore_backup.json:1]: remove `mmv1/third_party/validator/tests/data/example_filestore_backup.json` and `mmv1/third_party/validator/tests/data/example_filestore_backup.tf`, you don't need to have any validator related tests

File: mmv1/products/filestore/api.yaml [hunk 4]
```diff
@@ -327,3 +335,119 @@ objects:
...
+        description: |
+          Name of the file share in the source Cloud Filestore instance that the backup is created from.
+        input: true
+        required: true
+      - !ruby/object:Api::Type::Enum
+        name: 'sourceInstanceTier'
+        description: |
+          The service tier of the source Cloud Filestore instance that this backup is created from.
+        values:
+          - :TIER_UNSPECIFIED
```
Comment (c2thorn) [inline mmv1/products/filestore/api.yaml:436]:
```suggestion
```
we can omit this value, it is usually unusable

File: mmv1/products/filestore/api.yaml [hunk 5]
```diff
@@ -188,6 +188,14 @@ objects:
                       An integer representing the anonymous group id with a default value of 65534.
                       Anon_gid may only be set with squashMode of ROOT_SQUASH. An error will be returned
                       if this field is specified for other squashMode settings.
+            - !ruby/object:Api::Type::String
+              name: 'sourceBackup'
```
Comment (c2thorn) [inline mmv1/products/filestore/api.yaml:192]: This field is untested. Can you add a test for it?

**Review by c2thorn — CHANGES_REQUESTED** [review]

File: mmv1/products/filestore/api.yaml [hunk 6]
```diff
@@ -188,6 +194,7 @@ objects:
                       An integer representing the anonymous group id with a default value of 65534.
                       Anon_gid may only be set with squashMode of ROOT_SQUASH. An error will be returned
                       if this field is specified for other squashMode settings.
+              input: true
```
Comment (c2thorn) [inline mmv1/products/filestore/api.yaml:197]: why is this added to `nfsExportOptions` ?

File: mmv1/templates/terraform/examples/filestore_instance_full.tf.erb [hunk 7]
```diff
@@ -28,3 +28,15 @@ resource "google_filestore_instance" "<%= ctx[:primary_resource_id] %>" {
     connect_mode = "DIRECT_PEERING"
   }
 }
+
+resource "google_filestore_backup" "backup" {
```
Comment (c2thorn) [inline mmv1/templates/terraform/examples/filestore_instance_full.tf.erb:32]: This example's intention is to show the capabilities of Filestore instance, not Filestore Backup. We should remove changes to this example. From what I can tell, the backup part of this example doesn't test anything different than `mmv1/templates/terraform/examples/filestore_backup_basic.tf.erb`

File: mmv1/products/filestore/api.yaml [hunk 8]
```diff
@@ -144,6 +144,12 @@ objects:
                 File share capacity in GiB. This must be at least 1024 GiB
                 for the standard tier, or 2560 GiB for the premium tier.
               required: true
+            - !ruby/object:Api::Type::String
```
Comment (c2thorn) [inline mmv1/products/filestore/api.yaml:147]: Is adding this field part of the original request? We should not add it if not. Also, this is untested in this PR.

**Review by c2thorn — CHANGES_REQUESTED** [review]
Can you add the GH issue this PR will close? Like: fixes ___

File: mmv1/products/filestore/api.yaml (hunk 6 above)
Comment (googlerjk) [reply mmv1/products/filestore/api.yaml:197]: I got the latest and rebased, I didn't put this in there.

File: mmv1/products/filestore/api.yaml (hunk 8 above)
Comment (googlerjk) [reply mmv1/products/filestore/api.yaml:147]: It is an output field read onlly of that resource, I figured it should be supported with the backup implemented.

Comment (c2thorn) [reply mmv1/products/filestore/api.yaml:147]: Is it? You did not mark it as such.

File: mmv1/products/filestore/api.yaml (hunk 6 above)
Comment (c2thorn) [reply mmv1/products/filestore/api.yaml:197]: Looks like it was added here: https://github.com/GoogleCloudPlatform/magic-modules/pull/6742/commits/77d15ac22ba7617542cec16be41ec80d5f7d113b#diff-118310b4699d8c11e6cd16d34249e0e1a44f70fbc36f50373bfa1e15d604978dR198 

Seems like you meant it for the `sourceBackup` field, and it probably should have been `output` instead.

**Review by shuyama1 — DISMISSED** [review]
@googlerjk The latest commit does not seem to address all the comments @c2thorn left. Is this the latest version? Sorry if I have missed anything.
Comments that need to be addressed:
1. what's the purpose of modifying example at [mmv1/templates/terraform/examples/filestore_instance_full.tf.erb](https://github.com/GoogleCloudPlatform/magic-modules/pull/6742/files/6515acb233f76651569eba9bc379c5233ccc35bf#diff-ace55e9c3d3c953f02941d02cce5c760d163213e097d67482508796d363a9231)? The goal of that example is to test the `google_filestore_instance` resource
2. `input: true` is added to `nfsExportOptions` in `google_filestore_instance` resource? Is that intentional?

**Comment by googlerjk** [comment]
> @googlerjk The latest commit does not seem to address all the comments @c2thorn left. Is this the latest version? Sorry if I have missed anything. Comments that need to be addressed:
> 
> 1. what's the purpose of modifying example at [mmv1/templates/terraform/examples/filestore_instance_full.tf.erb](https://github.com/GoogleCloudPlatform/magic-modules/pull/6742/files/6515acb233f76651569eba9bc379c5233ccc35bf#diff-ace55e9c3d3c953f02941d02cce5c760d163213e097d67482508796d363a9231)? The goal of that example is to test the `google_filestore_instance` resource
> 2. `input: true` is added to `nfsExportOptions` in `google_filestore_instance` resource? Is that intentional?

RE: #1 since its the "full" version of the filstore, I included backup in it. And this is why I made the changes.
RE: #2 looks like my update didn't get pushed. The field should be set to input. I'll push that now.

File: mmv1/products/filestore/api.yaml (hunk 8 above)
Comment (shuyama1) [reply mmv1/products/filestore/api.yaml:147]: > > @googlerjk The latest commit does not seem to address all the comments @c2thorn left. Is this the latest version? Sorry if I have missed anything. Comments that need to be addressed:
> > 
> > 1. what's the purpose of modifying example at [mmv1/templates/terraform/examples/filestore_instance_full.tf.erb](https://github.com/GoogleCloudPlatform/magic-modules/pull/6742/files/6515acb233f76651569eba9bc379c5233ccc35bf#diff-ace55e9c3d3c953f02941d02cce5c760d163213e097d67482508796d363a9231)? The goal of that example is to test the `google_filestore_instance` resource
> > 2. `input: true` is added to `nfsExportOptions` in `google_filestore_instance` resource? Is that intentional?
> 
> RE: #1 since its the "full" version of the filstore, I included backup in it. And this is why I made the changes. RE: #2 looks like my update didn't get pushed. The field should be set to input. I'll push that now.

RE: #1
I would still suggest we remove the google_filestore_backup part from filestore_instance_full test, as the newly added part tests the exactly same functionality of the `google_filestore_backup` resource as in filestore_backup_basic test. It would not be necessary to test the same functionality in two tests.

**Comment by shuyama1** [comment]
> > @googlerjk The latest commit does not seem to address all the comments @c2thorn left. Is this the latest version? Sorry if I have missed anything. Comments that need to be addressed:
> > 
> > 1. what's the purpose of modifying example at [mmv1/templates/terraform/examples/filestore_instance_full.tf.erb](https://github.com/GoogleCloudPlatform/magic-modules/pull/6742/files/6515acb233f76651569eba9bc379c5233ccc35bf#diff-ace55e9c3d3c953f02941d02cce5c760d163213e097d67482508796d363a9231)? The goal of that example is to test the `google_filestore_instance` resource
> > 2. `input: true` is added to `nfsExportOptions` in `google_filestore_instance` resource? Is that intentional?
> 
> RE: #1 since its the "full" version of the filstore, I included backup in it. And this is why I made the changes. RE: #2 looks like my update didn't get pushed. The field should be set to input. I'll push that now.

RE: https://github.com/GoogleCloudPlatform/magic-modules/pull/1
I would still suggest we remove the google_filestore_backup part from filestore_instance_full test, as the newly added part tests the exactly same functionality of the google_filestore_backup resource as in filestore_backup_basic test. It would not be necessary to test the same functionality in two tests.

File: mmv1/products/filestore/api.yaml (hunk 8 above)
Comment (shuyama1) [reply mmv1/products/filestore/api.yaml:147]: I tested it locally and it doesn't seem the backup name is returned for this field. Filestore instance and backup got created successfully though. Do you happen to know why it's not got returned?

**Comment by shuyama1** [comment]
I tested it locally and it doesn't seem the backup name is returned for the field `source_backup` in `google_filestore_instance`. Filestore instance and backup got created successfully though. Do you happen to know why it's not got returned?

**Comment by googlerjk** [comment]
> I tested it locally and it doesn't seem the backup name is returned for the field `source_backup` in `google_filestore_instance`. Filestore instance and backup got created successfully though. Do you happen to know why it's not got returned?

No I dont know why its not returned, its stated to be returned in the documentation and it output only.

File: mmv1/products/filestore/api.yaml [hunk 9]
```diff
@@ -327,3 +334,110 @@ objects:
...
+        description: |
+          The amount of bytes needed to allocate a full copy of the snapshot content.
+        output: true
+      - !ruby/object:Api::Type::String
+        name: 'storageBytes'
+        description: |
+          The size of the storage used by the backup. As backups share storage, this number is expected to change with backup creation/deletion.
+        output: true
+      - !ruby/object:Api::Type::String
+        name: 'sourceInstance'
```
Comment (c2thorn) [inline mmv1/products/filestore/api.yaml:412]: There's no way this is updatable right? Shouldn't this sbe `input: true` ?

**Review by c2thorn — CHANGES_REQUESTED** [review]
My apologies for missing this up to this point, but we really need to add a test for updatability. Can you write a handwritten test to check the updatability of `description` and `labels` ?

**Comment by googlerjk** [comment]
> @googlerjk The latest commit does not seem to address all the comments @c2thorn left. Is this the latest version? Sorry if I have missed anything. Comments that need to be addressed:
> 
> 1. what's the purpose of modifying example at [mmv1/templates/terraform/examples/filestore_instance_full.tf.erb](https://github.com/GoogleCloudPlatform/magic-modules/pull/6742/files/6515acb233f76651569eba9bc379c5233ccc35bf#diff-ace55e9c3d3c953f02941d02cce5c760d163213e097d67482508796d363a9231)? The goal of that example is to test the `google_filestore_instance` resource
> 2. `input: true` is added to `nfsExportOptions` in `google_filestore_instance` resource? Is that intentional?

this has been applied.

File: mmv1/products/filestore/api.yaml (hunk 9 above)
Comment (googlerjk) [reply mmv1/products/filestore/api.yaml:412]: output only now

**Comment by googlerjk** [comment]
> My apologies for missing this up to this point, but we really need to add a test for updatability. Can you write a handwritten test to check the updatability of `description` and `labels` ?

Update made and pushed

**Review by c2thorn — APPROVED** [review]
